midp/media.js

```diff
@@ -385,6 +388,121 @@ ImagePlayer.prototype.setVisible = function(visible) {
 }
 
+function VideoPlayer(playerContainer) {
+    this.playerContainer = playerContainer;
+
+    this.video = document.createElement("video");
+    this.video.style.position = "absolute";
+    this.video.style.visibility = "hidden";
+
+    this.isVideoControlSupported = true;
+    this.isVolumeControlSupported = true;
+
+    // VideoPlayer::start could be called while the video element
+    // is hidden, causing the call to HTMLVideoElement::play to be
+    // ignored.
+    // Thus, we need to call HTMLVideoElement::play when the element
+    // gets visible.
+    this.isPlaying = false;
+}
+
+VideoPlayer.prototype.realize = function() {
+    var ctx = $.ctx;
+
+    var p = new Promise((function(resolve, reject) {
+        this.video.addEventListener("canplay", (function onCanPlay() {
+            this.video.removeEventListener("canplay", onCanPlay);
+            resolve(1);
+        }).bind(this));
+
+        this.video.onerror = function() {
+            ctx.setAsCurrentContext();
+            reject($.newMediaException("Failed to load video"));
+        };
+
+        if (this.playerContainer.url.startsWith("file")) {
+            fs.open(this.playerContainer.url.substring(7), (function(fd) {
+                var videoData = fs.read(fd);
+                fs.close(fd);
+                this.video.src = URL.createObjectURL(new Blob([ videoData ]),
+                                                     { type: this.playerContainer.contentType });
+            }).bind(this));
+        } else {
+            this.video.src = this.playerContainer.url;
+        }
+    }).bind(this));
+
+    p.done(function() {
+        if (!this.video.src) {
+            return;
+        }
+        URL.revokeObjectURL(this.video.src);
+    }.bind(this));
+
+    return p;
+}
+
+VideoPlayer.prototype.start = function() {
+    if (this.video.style.visibility === "hidden") {
+        this.isPlaying = true;
+    } else {
+        this.video.play();
+    }
+}
+
+VideoPlayer.prototype.pause = function() {
+    this.video.pause();
+    this.isPlaying = false;
+}
+
+VideoPlayer.prototype.close = function() {
+    if (this.video.parentNode) {
+        this.video.parentNode.removeChild(this.video);
+    }
+    this.pause();
+}
+
+VideoPlayer.prototype.getMediaTime = function() {
+    return Math.round(this.video.currentTime * 1000);
+}
+
+VideoPlayer.prototype.getWidth = function() {
+    return this.video.videoWidth;
+}
+
+VideoPlayer.prototype.getHeight = function() {
+    return this.video.videoHeight;
+}
+
+VideoPlayer.prototype.setLocation = function(x, y, w, h) {
+    this.video.style.left = x + "px";
+    this.video.style.top = y + "px";
+    this.video.style.width = w + "px";
+    this.video.style.height = h + "px";
+    document.getElementById("main").appendChild(this.video);
+}
+
+VideoPlayer.prototype.setVisible = function(visible) {
+    this.video.style.visibility = visible ? "visible" : "hidden";
+    if (this.isPlaying) {
```

Comment:
Done in 3bbdd3b (play is ignored if the element is hidden, so even without this change we wouldn't have started the video, but it's clearer this way).
There might be other things that we need to do when a MIDlet modifies the player's visibility several times, but it's hard to tell until we find a MIDlet that uses the video player in a more complicated way.